### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by SuicidalInsanity**
Bugfixes
-Fixes cost of aluminum armor panels
-Fueltanks will no longer randomly explode when shot
-Fixes ModulePartVariants giving negative money with wooden parts
-Fixes LeakFX not appearing
-Fixes Fuel leak duration
-Fix InertTanks/SelfSealing Tanks/Armored Cockpits not persisting through reloading
-AI can now select Lasers for attacking ground targets
Additions
-Weapons
 - Adds Chaingun functionality; chainguns have user configurable RoF in the SPH/VAB
 - Adds Weapon Spool up time functionality (for Gatling guns, etc)
 - Lasers get damage/FX growth option for increasing damage/changing beam width/color the longer they've been firing
 - Adds support for EMP, Beehive, Impulse, Gravitic, and Nuclear bullets
 - Expands GUI ammo labels to include above types    
 
-BattleDamage
 - Adds toggle if burning fueltanks explode
 - Toggling paintball Mode now toggles BattleDamage Off    
 
-Misc
 - Armor Panels now use Armor Integrity instead of HP
 - Nukes now support custom FX

### CODE REVIEW

File: BDArmory.Core/BDArmorySettings.cs [hunk 1]
```diff
@@ -43,9 +43,11 @@ public class BDArmorySettings
         [BDAPersistentSettingsField] public static bool VESSEL_RELATIVE_BULLET_CHECKS = false;
         [BDAPersistentSettingsField] public static bool AIM_ASSIST = true;
         [BDAPersistentSettingsField] public static bool DRAW_AIMERS = true;
-        [BDAPersistentSettingsField] public static bool DRAW_DEBUG_LINES = false;
+        // Debug Labels
+        [BDAPersistentSettingsField] public static bool DRAW_DEBUG_LINES = false;                   //AI/Weapon aim visualizers
         [BDAPersistentSettingsField] public static bool DRAW_DEBUG_LABELS = false;
         [BDAPersistentSettingsField] public static bool DRAW_ARMOR_LABELS = false;                 //armor only debug messages, for testing/debugging. remove/revert back to debug_labels later
+
```
Comment (SuicidalInsanity) [inline BDArmory.Core/BDArmorySettings.cs:50]: Was going to add new debug categories, but that's a bit outside of scope for this commit

File: BDArmory/Bullets/BulletInfo.cs [hunk 2]
```diff
@@ -26,17 +35,26 @@ public class BulletInfo
         public static HashSet<string> bulletNames;
         public static BulletInfo defaultBullet;
 
-        public BulletInfo(string name, float caliber, float bulletVelocity, float bulletMass,
-                          bool explosive, bool incendiary, float tntMass, string fuzeType, float apBulletDmg,
+        public static List<string> tagsList;
```
Comment (SuicidalInsanity) [inline BDArmory/Bullets/BulletInfo.cs:38]: Forgot to remove this (ln 38)

File: BDArmory/Bullets/BulletInfo.cs [hunk 3]
```diff
@@ -98,23 +122,30 @@ public static void Load()
                             Debug.LogError("[BDArmory.BulletInfo]: Bullet definition " + name_ + " from " + nodes[i].parent.name + " already exists, skipping.");
                         continue;
                     }
-                    Debug.Log("[BDArmory.BulletInfo]: Parsing definition of bullet " + name_ + " from " + nodes[i].parent.name);
```
Comment (SuicidalInsanity) [inline BDArmory/Bullets/BulletInfo.cs:101]: oops, that got accidentally removed

File: BDArmory/Bullets/PooledBullet.cs [hunk 4]
```diff
@@ -746,6 +795,16 @@ bool BulletHitAnalysis(RaycastHit hit, float period)
...
             }
+            if (EMP)
+            {
+                var emp = hitPart.vessel.rootPart.FindModuleImplementing<ModuleDrainEC>();
+                if (emp == null)
+                {
+                    emp = (ModuleDrainEC)hitPart.vessel.rootPart.AddModule("ModuleDrainEC");
+                } 
+                emp.incomingDamage += (caliber * Mathf.Clamp(bulletMass - tntMass, 0.1f, 101)); //soft EMP caps at 100; can always add a EMP amount value to bulletcfg later, but this should work for now
+                emp.softEMP = true;
```
Comment (SuicidalInsanity) [inline BDArmory/Bullets/PooledBullet.cs:806]: We may want to review bullet EMP damage at somepoint, but this should work for now

File: BDArmory/Bullets/PooledBullet.cs [hunk 5]
```diff
@@ -918,7 +977,7 @@ bool BulletHitAnalysis(RaycastHit hit, float period)
                 bool viableBullet = ProjectileUtils.CalculateBulletStatus(bulletMass, caliber, sabot);
 
                 ProjectileUtils.StealResources(hitPart, sourceVessel, stealResources);
-                ProjectileUtils.CheckPartForExplosion(hitPart);
+                //ProjectileUtils.CheckPartForExplosion(hitPart);
```
Comment (SuicidalInsanity) [inline BDArmory/Bullets/PooledBullet.cs:980]: This was the cause of fueltanks occasionally randomly exploding when shot, old legacy code

File: BDArmory/Modules/MissileFire.cs [hunk 6]
```diff
@@ -4171,6 +4171,64 @@ bool SmartPickWeapon_EngagementEnvelope(TargetInfo target)
                             }
                         }
 
+                        if (candidateClass == WeaponClasses.DefenseLaser) //lasers would be a suboptimal choice for strafing attacks, but if nothing else available...
```
Comment (BrettRyland) [inline BDArmory/Modules/MissileFire.cs:4174]: Didn't this get added already in 1.4.14.0 lines 4121—4172?

File: BDArmory/Distribution/GameData/BDArmory/BulletDefs/BD_Bullets.cfg [hunk 7]
```diff
@@ -276,6 +337,7 @@ BULLET
 BULLET
 {
     name = 25mmBullet
+	DisplayName = 25mm Bullet
```
Comment (SuicidalInsanity) [inline BDArmory/Distribution/GameData/BDArmory/BulletDefs/BD_Bullets.cfg:340]: some of these don't really give much to work with, for human-readable names

File: BDArmory/Modules/BDAdjustableArmor.cs [hunk 8]
```diff
@@ -325,6 +325,7 @@ void UpdateThickness()
             if (armor != null && armorTransforms != null)
             {
                 armorthickness = Mathf.Clamp((armor.Armor / 10), 0.1f, 1500);
+                if (isTriangularPanel && part.isMirrored) armorthickness = Mathf.Abs(armorthickness) * -1;
                 //if (!isCurvedPanel)
```
Comment (SuicidalInsanity) [inline BDArmory/Modules/BDAdjustableArmor.cs:329]: needs further work, but that will take more time if we want to get an update out in the near future

File: BDArmory/Modules/MissileFire.cs (hunk 6 above)
Comment (SuicidalInsanity) [reply BDArmory/Modules/MissileFire.cs:4174]: ok, didn't fully get dev build synced before working. That's quite the spot check fail. Yes, looks like I accidentally duplicated ground lasers

File: BDArmory.Core/BDArmorySettings.cs (hunk 1 above)
Comment (BrettRyland) [reply BDArmory.Core/BDArmorySettings.cs:50]: Yeah, that can be done later when there's less urgent things to do. It's a nice-to-have feature, but not necessary.

File: BDArmory.Core/Module/HitpointTracker.cs [hunk 9]
```diff
@@ -53,6 +53,8 @@ public class HitpointTracker : PartModule, IPartMassModifier, IPartCostModifier
         public float HullCostAdjust = 0f;
         double resourceCost = 0;
 
+        double resourceCost = 0;
```
Comment (BrettRyland) [inline BDArmory.Core/Module/HitpointTracker.cs:56]: This is a duplicate too.

Comment (SuicidalInsanity) [reply BDArmory.Core/Module/HitpointTracker.cs]: Oh, I know what's going on - when I originally pushed this, it was from partch-2, not dev, so loading the dev PR needed a fetch -> push, which resulted in a merge
fixed.

File: BDArmory/FX/FuelLeakFX.cs [hunk 10]
```diff
@@ -21,13 +22,37 @@ public static ObjectPool CreateLeakFXPool(string modelPath)
...
 
+        public float drainRate = 1;
+        private int fuelLeft = 0;
         public float lifeTime = 20;
         private float startTime;
         private float disableTime = -1;
         private float _highestEnergy = 1;
+
+        PartResource fuel;
+        PartResource ox;
```
Comment (BrettRyland) [inline BDArmory/FX/FuelLeakFX.cs:33]: Building this says that `FuelLeakFX.ox` and `FuelLeakFX.mp` are assigned but not used. You appear to be using local variables of the same name in `Update`.

**Pull request merged**